### ISSUE TITLE
0.16.0 - Occasionally vault init container just doesn't start

### ISSUE THREAD

**Issue by tim-sendible** [hashicorp/vault-k8s#349]
**Describe the bug**
I apologise in advance. This is admittedly a vague bug. But I'm hoping by raising it, others with the same issue can help add further information.

We had been running vault-k8s 0.14.2 in our clusters for some time. We extensively use vault init containers to inject secrets into our pods. This has worked fine through many thousands of invokations.

We have since upgraded to vault k8s 0.16.0, and very intermittently, the vault init container simply doesn't exist. The annotations are there, but the init container never runs... it doesn't even try to run. Instead, the main pod just starts, and then crashes because the expected secrets aren't present. This issue occurs maybe 1% of the time.

Shooting the pod is often enough to make it work a second time. This makes the issue very tricky to troubleshoot. There was one occasion where I had to shoot the vault-agent-injector pods themselves before the problem would resolve itself.

At the time of this error happening, the vault-agent-injector pods seem to be running fine. They haven't just restarted or anything similar.

We cannot find any logs that offer anything relevant I'm afraid.

Reverting vault-k8s to 0.14.2 causes the problem to go away.

**To Reproduce**
1. Deploy application annotated for vault-agent injection
2. Occasionally, the pod does not start with the vault-agent init container. Shooting the pod and trying again then works.

**Expected behavior**
The vault agent injection init container should always run on appropriately annotated pods.

**Environment**
* Kubernetes version: 
  * Distribution or cloud vendor (OpenShift, EKS, GKE, AKS, etc.): EKS 1.22.6
  * Other configuration options or runtime services (istio, etc.): none
* vault-k8s version: 0.16.0. Vault 1.10.3

**Additional context**

**Comment by matsgoran** [comment]
We are experiencing the same issue although in an on-premise cluster. 

We have been running vault (via helm 0.19.0 and earlier) for months without any issues. 

After upgrading to 0.20.0 a few days ago vault container injection has started to fail. Killing the vault injector pods fixes the issue temporarily (24h). 

**Environment:**
Kubernetes version: v1.20.7 (on-prem)
Vault: 
- helm chart version: 0.20.0
- vault 1.10.3 in HA-mode (5 nodes)
- vault-k8s 0.16.0 (same amount of injector instances as vault)

After inspecting the logs this may be related certificate auto-rotation. Without any knowledge of the innards of vault-k8s I suspect that this may have something to do with #336 introduced in 0.16.0.

Container Injection stops working and we get bad certificate errors in the vault-k8s logs:
```
Using internal leader elector logic for webhook certificate management
Registering telemetry path on "/metrics"
{"@level":"info","@message":"Starting handler..","@module":"handler","@timestamp":"2022-05-20T13:59:49.850133Z"}
Listening on ":8080"...
{"@level":"info","@message":"Updated certificate bundle received. Updating certs...","@module":"handler.certwatcher","@timestamp":"2022-05-20T13:59:49.952398Z"}
{"@level":"info","@message":"Webhooks changed. Updating certs...","@module":"handler.certwatcher","@timestamp":"2022-05-20T13:59:49.952942Z"}
{"@level":"info","@message":"Webhooks changed. Updating certs...","@module":"handler.certwatcher","@timestamp":"2022-05-20T13:59:49.953041Z"}
{"@level":"info","@message":"Webhooks changed. Updating certs...","@module":"handler.certwatcher","@timestamp":"2022-05-20T13:59:49.953128Z"}
{"@level":"info","@message":"Currently the leader","@module":"handler.auto-tls","@timestamp":"2022-05-20T13:59:54.851917Z"}
{"@level":"info","@message":"Generated CA","@module":"handler.auto-tls","@timestamp":"2022-05-20T13:59:54.852982Z"}
{"@level":"info","@message":"Updated certificate bundle received. Updating certs...","@module":"handler.certwatcher","@timestamp":"2022-05-20T13:59:54.874522Z"}
{"@level":"info","@message":"Webhooks changed. Updating certs...","@module":"handler.certwatcher","@timestamp":"2022-05-20T13:59:54.885696Z"}
{"@level":"info","@message":"Currently the leader","@module":"handler.auto-tls","@timestamp":"2022-05-20T13:59:59.851417Z"}
{"@level":"info","@message":"Request received","@module":"handler","@timestamp":"2022-05-20T21:00:10.379849Z","Method":"POST","URL":{"Scheme":"","Opaque":"","User":null,"Host":"","Path":"/mutate","RawPath":"","ForceQuery":false,"RawQuery":"timeout=30s","Fragment":"","RawFragment":""}}
{"@level":"info","@message":"Request received","@module":"handler","@timestamp":"2022-05-21T03:00:07.186572Z","Method":"POST","URL":{"Scheme":"","Opaque":"","User":null,"Host":"","Path":"/mutate","RawPath":"","ForceQuery":false,"RawQuery":"timeout=30s","Fragment":"","RawFragment":""}}
{"@level":"info","@message":"Request received","@module":"handler","@timestamp":"2022-05-21T08:00:09.706560Z","Method":"POST","URL":{"Scheme":"","Opaque":"","User":null,"Host":"","Path":"/mutate","RawPath":"","ForceQuery":false,"RawQuery":"timeout=30s","Fragment":"","RawFragment":""}}
{"@level":"info","@message":"Request received","@module":"handler","@timestamp":"2022-05-21T08:00:10.478049Z","Method":"POST","URL":{"Scheme":"","Opaque":"","User":null,"Host":"","Path":"/mutate","RawPath":"","ForceQuery":false,"RawQuery":"timeout=30s","Fragment":"","RawFragment":""}}
{"@level":"info","@message":"Request received","@module":"handler","@timestamp":"2022-05-21T08:00:10.716350Z","Method":"POST","URL":{"Scheme":"","Opaque":"","User":null,"Host":"","Path":"/mutate","RawPath":"","ForceQuery":false,"RawQuery":"timeout=30s","Fragment":"","RawFragment":""}}
{"@level":"error","@message":"http: TLS handshake error from 10.233.104.0:22021: remote error: tls: bad certificate","@module":"handler","@timestamp":"2022-05-21T20:00:09.062720Z"}
{"@level":"error","@message":"http: TLS handshake error from 10.233.123.0:64406: remote error: tls: bad certificate","@module":"handler","@timestamp":"2022-05-21T23:00:03.095288Z"}
{"@level":"error","@message":"http: TLS handshake error from 10.233.123.0:1774: remote error: tls: bad certificate","@module":"handler","@timestamp":"2022-05-22T04:00:09.291275Z"}
.
.
.
``` 

Snippet from the injector configuration from our helm deployment
```yaml
.
.
.
injector:                                                                                                                                                                                                          
  # True if you want to enable vault agent injection.                                                                                                                                                              
  # @default: global.enabled                                                                                                                                                                                       
  enabled: "-"                                                                                           

  replicas: 5

  # Configures the port the injector should listen on
  port: 8080

  # If multiple replicas are specified, by default a leader will be determined
  # so that only one injector attempts to create TLS certificates.
  leaderElector:
    enabled: true

  # If true, will enable a node exporter metrics endpoint at /metrics.
  metrics:
    enabled: true

  # External vault server address for the injector to use. Setting this will
  # disable deployment of a vault server along with the injector.
  externalVaultAddr: ""

  # image sets the repo and tag of the vault-k8s image to use for the injector.
  image:
    repository: "hashicorp/vault-k8s"
    #There seems to be a bug in 0.16.0 which leads to the auto tls update feature failing on next certificate
    # rotation. (after 24 hours). Follow upstream when issue is resolved
    tag: "0.16.0"
    pullPolicy: IfNotPresent

  # agentImage sets the repo and tag of the Vault image to use for the Vault Agent
  # containers.  This should be set to the official Vault image.  Vault 1.3.1+ is
  # required.
  agentImage:
    repository: "hashicorp/vault"
    tag: "1.10.3"

  # The default values for the injected Vault Agent containers.
  agentDefaults:
    # For more information on configuring resources, see the K8s documentation:
    # https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
    cpuLimit: "500m"
    cpuRequest: "250m"
    memLimit: "128Mi"
    memRequest: "64Mi"

    # Default template type for secrets when no custom template is specified.
    # Possible values include: "json" and "map".
    template: "map"

   # Default values within Agent's template_config stanza.                                              
    templateConfig:                                                                                      
      exitOnRetryFailure: true                                                                                                                                                                                     
      staticSecretRenderInterval: ""                                                                     
                                                                                                         
  # Mount Path of the Vault Kubernetes Auth Method.
  authPath: "auth/kubernetes"

  # Configures the log verbosity of the injector.
  # Supported log levels include: trace, debug, info, warn, error
  logLevel: "info"

  # Configures the log format of the injector. Supported log formats: "standard", "json".
  logFormat: "json"

  # Configures all Vault Agent sidecars to revoke their token when shutting down
  revokeOnShutdown: false
.
.
.
```

We have temporarily downgraded vault-k8s to 0.15.0 and will know for sure tomorrow morning if this solves our issues.

**Comment by matsgoran** [comment]
#275

**Comment by tvoran** [comment]
Hi folks, thanks for the report and the digging here. This should be fixed in v0.16.1 which includes this fix: https://github.com/hashicorp/vault-k8s/pull/350

Closing for now.
